Keychain and settings

noamkuMoDev opened the pull request:
1. Fixed notes from previous PR (turned strings to static vars, changed name of function that returns boolean)

2. Added AppSettings file in "Models" folder that is a singelton for the app settings based on what's saved inside UserDefaults. Fetches and changes to the app setting are performed through it (get & set for each setting)

3. Added KeychainManager file in "Persistence" folder with fetch, add, remove functions.

4. To check if user is logged in - check if email exists in keychain. If exists - check is user is logged in on firebase and display a screen accordingly (login or homepage). If not - load Login screen.

noamkuMoDev committed:
Added KeychainManager

noamkuMoDev committed:
Added AppSettings singleton class, Added check for user email in Keychain before Firebase auth check
Fixed notes from previous PR (static vars, boolean func name)

noamkuMoDev committed:
Writing, fetching and removing FavoriteArticle objects from CoreData is working

noamkuMoDev committed:
Used notificationsCenter to refresh tableViews on other VCs
Turned newsArray to singleton to access in on different instances of BaseViewModel

noamkuMoDev committed:
Re-ordered the favorites articles display order

abenartz commented on Dispatcher/UI/Profile/Settings/SettingsViewController.swift:
```diff
@@ -44,11 +44,17 @@ extension SettingsViewController: UITableViewDataSource {
         let cell = tableView.dequeueReusableCell(withIdentifier: Constants.TableCellsIdentifier.SETTING, for: indexPath) as! AppSettingCell
-        
+        cell.delegate = self
+        print("======")
```
We don't keep this kind of logs in the PR. Please remove unnecessary logs.

noamkuMoDev replied:
Fixed
@abenartz

abenartz commented on Dispatcher/UI/Profile/Settings/SettingsViewModel.swift:
```diff
@@ -0,0 +1,49 @@
+    ]
+    
+    func fetchUserSettingsPreferences() {
+        appSettings[0].options[0].status = repository.fetchSaveFiltersSetting()
```
fetching is usually refers to get remote data (from server..).
I would change the name to 'getSaveFiltersStatus'.

same with the others below.

noamkuMoDev replied:
Fixed in all places
@abenartz

abenartz commented on Dispatcher/UI/Profile/Settings/SettingsViewModel.swift:
```diff
@@ -0,0 +1,49 @@
+    func updateSetting(settingTitle: String, completionHandler: @escaping () -> ()) {
+        var modelIndex = -1
+        var settingIndex = -1
+        for (i, settingModel) in appSettings.enumerated() {
```
We don't want to get situation of O(n^2) - loop inside loop. this is bad for runtime performance.
Usually we can fix it using different data structure as Dictionary (hashmap).

noamkuMoDev committed:
Fixed function names (fetch ---> get)
Removed unnecessary comments and logs

noamkuMoDev committed:
Updated functionality flows
Removed unnecessary singletons

noamkuMoDev committed:
Updated flow logics and data refreshes

noamkuMoDev committed:
Turned appSettings array into Dictionary
Turned functions to more generic, renamed variables and functions, turned code to be more readable

noamkuMoDev committed:
Added timestamp property to SavedArticles (CoreData and Firestore)
Reviewed ALL functions and confirmed them
